qiskit/circuit/parameterexpression.py

```diff
@@ -255,6 +272,9 @@ def __int__(self):
                             'cannot be cast to an int.'.format(self.parameters))
         return int(self._symbol_expr)
 
+    def __hash__(self):
+        return hash((tuple(hash(p) for p in self.parameters), str(self)))
```

Comment:
```suggestion
        return hash((tuple(sorted(self.parameters, key=lambda p: p.name)), str(self)))
```
Iteration order isn't guaranteed for `self.parameters`, which could lead to some hard to track down bugs. ( This is a regular pain point of the parameters API, see https://github.com/Qiskit/qiskit-terra/pull/3910 ).

Reply:
Out of curiosity, where is `ParameterExpression.__hash__` being used? (`ParameterExpression.__eq__` was added for testing the equivalence library, so there might be some edge cases/performance questions on using `srepr` as the key of a large dict.)

Reply:
thank you